server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -36,6 +40,23 @@
 
     public static final String BACKING_INDEX_PREFIX = ".ds-";
     public static final DateFormatter DATE_FORMATTER = DateFormatter.forPattern("uuuu.MM.dd");
+    // Datastreams' leaf readers should be sorted by desc order of their timestamp field, as it allows search time optimizations
+    public static Comparator<LeafReader> DATASTREAM_LEAF_READERS_SORTER =
+        Comparator.comparingLong(
+            (LeafReader r) -> {
+                try {
+                    PointValues points = r.getPointValues(DataStream.TimestampField.FIXED_TIMESTAMP_FIELD);
+                    if (points != null) {
+                        byte[] sortValue = points.getMaxPackedValue();
+                        return LongPoint.decodeDimension(sortValue, 0);
+                    }
+                } catch (IOException e) {
+                    assert false : "Datastream index segment doesn't contain an expected " +
+                        DataStream.TimestampField.FIXED_TIMESTAMP_FIELD + " field!";
+                }
+                // this should not happen, as all data stream segments must contain @timestamp field
```

Comment:
should we have an assert here too ?

Reply:
@jimczi  I've opted to throw IllegalStateException instead of assertion. This allows us to to return an arbitrary value as it was before.

I hope you are fine with the change.